cli/flox-rust-sdk/src/providers/catalog.rs

```diff
@@ -1699,12 +1703,115 @@ impl BaseCatalogInfo {
     }
 }
 
+/// Derive the nixpkgs url to be used for builds.
+/// If a stability is provided, try to retrieve a url for that stability from the catalog.
+/// Else, if we can derive a stability from the toplevel group of the environment, use that.
+/// Otherwise attrr
+pub async fn base_catalog_url_for_stability_arg(
+    stability: Option<&str>,
+    base_catalog_info_fut: impl IntoFuture<Output = Result<BaseCatalogInfo, CatalogClientError>>,
+    toplevel_derived_url: Option<&BaseCatalogUrl>,
+) -> Result<BaseCatalogUrl, CatalogClientError> {
+    let url = match (stability, toplevel_derived_url) {
+        (Some(stability), _) => {
+            let base_catalog_info = base_catalog_info_fut.await?;
+            let make_error_message = || {
+                let available_stabilities = base_catalog_info.available_stabilities().join(", ");
+                formatdoc! {"
+                    Stability '{stability}' does not exist (or has not yet been populated).
+                    Available stabilities are: {available_stabilities}
+                "}
+            };
+
+            let url = base_catalog_info
+                .url_for_latest_page_with_stability(stability)
+                .ok_or_else(|| CatalogClientError::StabilityError(make_error_message()))?;
```

Comment:
Reviewer note: making this a `CatalogClientError` instead of an `anyhow::Error` was the only change I made when copying this from it's previous location, and was necessary because we don't use `anyhow` in `flox-rust-sdk`.